deep-learning/_posts/2017-02-09-batch-norm.md

```diff
@@ -0,0 +1,103 @@
+    text-align: justify; /* helps the reading flow */
+  }
+</style>
+
+# Why use it
+
+We want to reduce _**Internal Covariate Shift**_, the change in the distributions of internal nodes of a deep network during training. It is advantageous for the distribution of a layer input to remain fixed over time. To do so, we will **fix the means and variances of layer inputs**.
+
+Consider a layer with sigmoid activation $$ z = g(Wu + b) $$. For all dimensions of $$ x = Wu + b $$ except those with small absolute values, the gradient flowing down to $$ u $$ will vanish and the model will train slowly. Since $$ x $$ is affected by $$ W, b $$ and the parameters of the layers below, changes to those parameters during training will likely move many dimensions of $$ x $$ into the saturated regime of the sigmoid and **slow down training**.
+
+Training converges faster when the inputs of the network are whitened (linearly transformed to have $$ \mu = 0, \sigma^2 = 1 $$, and decorrelated). Thus, **we'd like to apply whitening to the input of each layer**.
+
+# How it works
+
+**Normalize each feature independently.**{: .phead} Consider a layer with $$ d $$-dimensional input $$ \bold{x} = (x^{(1)} \ldots x^{(d)}) $$. We normalize each dimension $$ k $$ like so :
+
+$$ \hat{x}^{(k)} = \frac{x^{(k)} - \mathrm{E}[x^{(k)}]}{\sqrt{\mathrm{Var}[x^{(k)}]}} $$
+
+where the mean and variance are computed on the *mini-batch*.
+
+**Add an affine transform just after.**{: .phead} Only applying a normalization can reduce the representational power of the network. For example, a sigmoid would be constrained to its linear regime. Following the normalization by a learnable affine transform addresses this issue. The BN operation becomes :
+
+$$ y^{(k)} = \gamma^{(k)}\hat{x}^{(k)} + \beta^{(k)} $$
+
+Notice that if the parameters would learn the values $$ \gamma^{(k)} = \sqrt{\mathrm{Var}[x^{(k)}]} $$ and $$ \beta^{(k)} = \mathrm{E}[x^{(k)}] $$, we would recover the original activations, if that were the optimal thing to do.
+
+**Constant normalization during inference.**{: .phead} During inference, we want the output to depend only on its input (not on other examples). For this, instead of the mini-batch statistics, we use statistics from the whole population, which are constant. In practice, we compute these statistics during training using moving averages. Since the normalization is now fixed, it can be composed with the scaling by $$ \gamma $$ and shift by $$ \beta $$, to yield a single linear transform.
+
+**Why do we need to backprop through it ?**{: .phead} Consider a layer $$ x = u + b $$, where $$ b $$ is a learned bias. We normalize the result using the mean of the activation computed over the whole training data : $$ \hat{x} = x - \mathrm{E}[x] $$. If a gradient descent step ignores the dependence of $$ \mathrm{E}[x] $$ on $$ b $$, then it will update $$ b \leftarrow b + \Delta b $$, where $$ \Delta b \propto \partial \text{loss} / \partial \hat{x} $$. Then,
```

Comment:
$$ \Delta b \propto \partial \text{loss} / \partial \hat{x}

Not shown properly also loss is a function which takes x (therefore the derivative on x)